### ISSUE TITLE
Hugo, front matter title

### ISSUE THREAD

**Issue by illume** [tgreyuk/typedoc-plugin-markdown#197]
Hello,

I'd like to add the title in front-matter for using the markdown in Hugo.

I guess replacing the header/title like this:
```
# Module: lib/someLib
```

With this:

```
---
title: Module lib/someLib
---
```

Is there perhaps a way to do this with an option, or even a theme?

Thanks for any pointers in the right direction.

**Comment by tgreyuk** [comment]
@illume  I think providing some kind of front matter support would be useful. We can provide a new option `front-matter` or `front-matter-title` that could do this quite easily. The question is would there be a need to add any other variables added to the front matter? (although we are limited as to what can be inferred).

The other thing to consider is how to handle main title visibility - just hide it by default or provide another option to hide it. Maybe just hide it to begin with so we don't have too many options.

It is also possible to write a custom markdown theme - some pointers here https://github.com/tgreyuk/typedoc-plugin-markdown/tree/master/packages/typedoc-github-wiki-theme/src although may not be necessary for this..

**Comment by illume** [comment]
Cool, thanks Tom for the tip about the theme.

I think `title:` is probably the main thing.

Other ones like slug could be useful. For example...

```
# Module: lib/someLib
```

With this:

```
---
title: Module lib/someLib
slug: lib/someLib
---
```

Maybe also `linkTitle` would be useful ("*used for creating shorter links to content if the title is too long; if set, Hugo defaults to using the linktitle before the title. Hugo can also order lists of content by linktitle*"), but I'm not sure how that could be derived automatically.

Here's the full list of things Hugo supports: https://gohugo.io/content-management/front-matter/
I see some other systems use front matter too (the docosaurus plugin and the vue one).

For basic integration with Hugo, I think only the title front-matter is required.

**Comment by illume** [comment]
> The other thing to consider is how to handle main title visibility - just hide it by default or provide another option to hide it. Maybe just hide it to begin with so we don't have too many options.

Yeah, I don't know what is best here. From the perspective of Hugo, hiding it is needed. Yeah, avoiding options if possible sounds like a good plan.

**Comment by tgreyuk** [comment]
@illume thanks for the comments. 

Having considered this in more detail I think extending the theme is probably the cleanest solution and gives us the greatest flexibility going forward.  This is what it looks like on the branch i am working on - https://github.com/tgreyuk/typedoc-plugin-markdown/blob/hugo-theme/packages/typedoc-hugo-theme/src/theme.ts. Its pretty simple - we can start with just the title but can extend it as much as we like. Could also do some stuff beyond front matter if required.

Just to note that we have found it safest to double quote string values:

```
---
title: "Class: SomeClass"
---
```

The usage would be:

```
typedoc --theme ./node_modules/typedoc-hugo-theme/dist ...options
```

If this looks like something you could work with I will publish it out so we can start playing with it?

**Comment by illume** [comment]
Looks good to me :) Thanks.

I'll try it out.

**Comment by tgreyuk** [comment]
cool - just making a few tweaks - will let you know when ready....

**Comment by tgreyuk** [comment]
Initial release -https://github.com/tgreyuk/typedoc-plugin-markdown/tree/master/packages/typedoc-hugo-theme ..

Have a play with it and see what you think.

Couple of initial thoughts

- Should the entry page be `index.md` rather than `README.md`?
- Should the slug have an option to add a base path such as `/api`?

**Comment by illume** [comment]
Cool.

I tried it out, and it generates all the front matter. To your questions...

> Should the entry page be index.md rather than README.md?
It uses "_index.md" as an index page it seems, and doesn't seem to pick up the README.md as an index file.

> Should the slug have an option to add a base path such as /api?

I'm putting the files into an docs/api folder, and I think people can do this. I'm not sure if it's also useful or not to have that option.

It seems to use the package.json name for the entry page title. Maybe this part could be an option. Maybe the title for the entry page could be just "API" or "`${packageName} API`" for this?

There's a few weird things going on. I'll have to reduce the data and do some more debugging find out exactly what's causing these.

- The URLs seem to be doubling up on part of the URL. eg. `modules/modules/lib_somelib/`. The "modules" part is repeated (and interfaces/interfaces).
- The links in the sidebar are all going together. Rather than Overview, modules, interfaces, and those having submenus - it puts all the modules, interfaces, and such all together.

I'll try and remove some files and do some debugging, and get back to you.

**Comment by illume** [comment]
For the slug has extra "module" and extra "interfaces" in the urls issue, 
I think it should only use the last part of page.url.

Example: `modules/components_common.md` the slug should be this:

```yaml
---
title: "Module: components/common"
slug: "components_common"
linkTitle: "components/common"
---
```

Currently there's an extra "modules/" the start of slug.

Something like this should make it use the last part of the slug:
```typescript
    onHugoPageEnd(page) {
        // yaml variables
        const slug = page.url.replace(/\.[^.$]+$/, '');
        // Only use the last part of the slug, because the folder is used to make URL
        let slugEnd = slug.split("/").pop();
        if (slugEnd === undefined) {
            slugEnd = slug;
        }
        const yamlVars = {
            title: front_matter_1.getPageTitle(page),
            slug: slugEnd,
            linkTitle: page.model.name,
        };
        // pass yaml variables to prependYAML util
        if (page.contents) {
            page.contents = front_matter_1.prependYAML(page.contents, yamlVars);
        }
    }

```

**Comment by tgreyuk** [comment]
ok - this should be available in typedoc-hugo-theme@0.1.1

**Comment by tgreyuk** [comment]
> It seems to use the package.json name for the entry page title. Maybe this part could be an option. Maybe the title for the entry page could be just "API" or "${packageName} API" for this?

Try  `--indexTitle API `

**Comment by illume** [comment]
> ok - this should be available in typedoc-hugo-theme@0.1.1

Cool, the slugs are working.

> Try --indexTitle API 

With this the modules.md has:

```yaml
---
title: "API"
slug: "README"
linkTitle: "packagenameturnsuphere"
---
```
Note the linkTitle here is still the package.json `"name"`.

If adding the option `"readme": "none",` it makes modules.md turn up as README.md.

**Comment by illume** [comment]
Cool, I figured out the sidebar issue :)

To fix the crowding of the side bar with every interface and module (and have "Modules", and "Interfaces" both appear in the side bar) these two index files should be generated. Without them Hugo decides to put every subfolder item together at the top level in the side bar.

modules/_index.md
```yaml
---
title: "Modules"
---
```

interfaces/_index.md
```yaml
---
title: "Interfaces"
---
```

**Comment by tgreyuk** [comment]
new release 0.2.0

- As per your previous comments entryDocument is defaulted to `_index.md`.
- Have fixed issues with `linkTitle` for index/modules files and readme options (i think)
- Have created the root `_index.md` for each folder group to solve the grouping issue.

The root slug is now `slug: "_index"`. Not sure if this is correct - maybe default slug should be `'/'` or something?

**Comment by illume** [comment]
Hello! It's working nicely for me :) 

Many thanks :tada: :balloon: 

A few notes...

The root slug doesn't seem to have an affect, so I guess it could be not generated at all. But doesn't seem to have any harm to me just leaving it as is.

I couldn't get entryDocument working, but decided not to use it anyway. I tried "entryDocument": "src/API.md", and on the command line with `npx typedoc --entryDocument src/API.md`. It seemed to use the README.md instead unless I used `"readme": "none",` in the tsconfig.

This is the config I ended up using (for any internet passers by).
```
  "typedocOptions": {
    "out": "../docs/development/api",
    "entryPoints": ["src/lib/k8s/index.ts", "src/lib/util.ts"],
    "theme": "./node_modules/typedoc-hugo-theme/dist",
    "excludeExternals": true,
    "excludePrivate": true,
    "excludeProtected": true,
    "hideBreadcrumbs": true,
    "readme": "none",
    "indexTitle": "API"
  }
```
